### PR TITLE
feat: Improve mapping functionality

### PR DESCRIPTION
**Pull request by Nurkambay**
This improvement is needed to properly map possible dialect's statement constructions like:
```
STATEMENT VAR1 BY VAR2 WITH PAR1 AND PAR2
``` 
that can be replaced by the dialect to the code:

```
MOVE  VAR1 TO VAR2.
PERFORM PAR1 THRU PAR2.
```

For now it is not possible to map VAR1, PAR1 and PAR2 to the original VAR1, PAR1 and PAR2 and in the same time all other characters to the ```STATEMENT``` keyword.

This PR is covering this case

## Checklist:
- [x] Each of my commits contains one meaningful change
- [x] I have performed rebase of my branch on top of the development
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  private void scanForTokens(
+      Map<String, Token> tokens, String[] statementMapArray, int index, Range range) {
+    char[] statementLine = statementMapArray[index].toCharArray();
+
+    int symbolCount = 0;
+    StringBuilder temp = new StringBuilder();
+
+    for (int i = 0; i < statementLine.length; i++) {
+      if (symbolCount % 2 == 0) {
+        if (statementLine[i] == '{') {
```
Comment (slavek-kucera) [inline server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java]: Is there an escape sequence when the `{` and `}` are actually part of the text?

Comment (Nurkambay) [reply server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java]: I cannot imagine when spaces can be needed here, so we can skip them (if needed)

Comment (Nurkambay) [reply server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java]: For now spaces are considered as a part of the token name

Comment (slavek-kucera) [reply server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java]: I am talking about braces not spaces.

Comment (Nurkambay) [reply server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java]: It is not a text, it is a map, and the map can be build without any text at all, using only spaces and braces with token names inside.

Comment (Nurkambay) [reply server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java]: Well, it is possible only when user need to replace statement with the text that contains braces. In that case the braced will be processed as a token name.

File: server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/SingleLineReplaceStrategy.java [hunk 2]
```diff
@@ -43,7 +43,7 @@ private void addLines(
       extendedText.insert(
           new Position(position.getLine() + 1, 0),
           new ExtendedTextLine(newLines[newLines.length - 1], instantLocation));
-      for (int i = 1; i < newLines.length - 2; i++) {
+      for (int i = 1; i < newLines.length - 1; i++) {
```
Comment (Nurkambay) [inline server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/SingleLineReplaceStrategy.java:46]: Old bug. I have added a test for it

File: server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java [hunk 3]
```diff
@@ -373,4 +373,289 @@ void testInsertWithPadding() {
     assertEquals(
         expectedCopybook.getLines().get(2).toString(), copybook1.getLines().get(2).toString());
   }
+
+  @Test
+  void testReplace_three_lines() {
```
Comment (Nurkambay) [inline server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java:380]: Test for the old bug

File: server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+    tokenReplacements.forEach(
+        (r, t) -> {
+          extendedText.delete(r);
+          extendedText.insert(
+              r.getStart(), new ExtendedTextLine(t.getValue(), t.getOriginalLocation()));
+        });
+  }
+
+  private void scanForTokens(
+      Map<String, Token> tokens, String[] statementMapArray, int mapLine, Range range) {
```
Comment (slavek-kucera) [inline server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java]:
```suggestion
  private void scanForTokens(
      Map<String, Token> tokens, char[] statementLine, int mapLine, int line) {
```

File: server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+                    new Position(line, character),
+                    new Position(line, character + tokenName.length() - 1)));
+        ExtendedTextLine extendedTextLine = extendedText.getLines().get(line);
+        String value =
+            extendedTextLine.subline(character, character + tokenName.length() - 1).toString();
+
+        tokens.put(tokenName, new Token(value, originalLocation));
+
+        symbolCount++;
+        temp = new StringBuilder();
```
Comment (slavek-kucera) [inline server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java]:
```suggestion
        temp.setLength(0);
```

File: server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+      Position startPosition,
+      Map<String, Token> tokens,
+      String replacementMap,
+      Map<Range, Token> tokenReplacements) {
+    StringBuilder result = new StringBuilder();
+    String[] replacementMapArray = MappingHelper.split(replacementMap);
+
+    for (int mapLine = 0; mapLine < replacementMapArray.length; mapLine++) {
+      char[] replacementLine = replacementMapArray[mapLine].toCharArray();
+      int bracesIndicator = 0;
```
Comment (slavek-kucera) [inline server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java]: Does it make sense to have the replacement span multiple lines?

File: server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+          tokenName.append(c);
+        } else {
+          outputLine.append(c);
+        }
+      }
+      if (mapLine != replacementMapArray.length - 1) {
+        outputLine.append("\n");
+      }
+      result.append(outputLine);
+      if (bracesIndicator > 0) {
```
Comment (slavek-kucera) [inline server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java:197]: check also for pending escape sequences?

File: server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+          String value = Optional.ofNullable(tokenNameAndValue.getRight()).orElse(token.getValue());
+          int character = outputLine.length() + (mapLine == 0 ? startPosition.getCharacter() : 0);
+          Range range =
+              new Range(
+                  new Position(startPosition.getLine() + mapLine, character),
+                  new Position(startPosition.getLine() + mapLine, character + value.length()));
+          tokenReplacements.put(range, new Token(value, token.getOriginalLocation()));
+          outputLine.append(value);
+
+          tokenName = new StringBuilder();
```
Comment (slavek-kucera) [inline server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java]:
```suggestion
          tokenWithValue.setLength(0);
```

File: server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/MappingHelperTest.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+            () -> MappingHelper.validateRange(new Range(new Position(5, 0), new Position(4, 0))));
+    assertEquals("Invalid range", exception.getMessage());
+  }
+
+  @Test
+  void testValidateRange_character_start_greater_character_end() {
+    Exception exception =
+        assertThrowsExactly(
+            IllegalArgumentException.class,
+            () -> MappingHelper.validateRange(new Range(new Position(0, 7), new Position(4, 2))));
```
Comment (slavek-kucera) [inline server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/MappingHelperTest.java]: This range seems valid. It starts on line 0 and ends on line 4.

Comment (Nurkambay) [reply server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/MappingHelperTest.java]: Fixed + test added

File: server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java [hunk 10]
```diff
@@ -301,13 +235,37 @@ private void validateMapSize(String map) {
...
+    List<ExtendedTextLine> lines = extendedText.getLines();
+    if (range.getStart().getLine() >= lines.size()) {
+      throw new IllegalArgumentException(entityName + "range start line out of bounds");
+    }
+    if (range.getStart().getCharacter() > lines.get(range.getStart().getLine()).size()) {
+      throw new IllegalArgumentException(entityName + "range start character out of bounds");
+    }
+    if (range.getEnd().getLine() >= lines.size()) {
+      throw new IllegalArgumentException(entityName + "range end line out of bounds");
+    }
```
Comment (slavek-kucera) [inline server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java:249]: I think the equality might actually be valid if the last character in the text was newline.

File: server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+ * of the token, situated in the appropriate range. For example: text document contains the line:
+ * "this and this text will be replaced" (*). The statement map can be: "{this} and {that}" (1).
+ * {this} points to 1st "this" token and {that} points to the 2nd "this" token. These token names:
+ * {this} and {that} now can be replaced with the replacement map. Replacement map can be: "{this}
+ * and even {that} text was replaced" (2). When TextMapReplacer apply replacement using (1)
+ * statement map and replacement map (2) to the document line "this and this text will be replaced"
+ * the result will be: "this and even that text was replaced", so token name {that} was substituted
+ * with actual token value "this". The mapping location for the 1st "this" token will be pointing to
+ * the 1st original "this" token and for the 2nd "this" token mapping will be pointing to the 2nd
+ * original "this" token. All other characters will be pointed to the provided range
```
Comment (slavek-kucera) [inline server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java:41]: This should be updated.

File: server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java (hunk 10 above)
Comment (Nurkambay) [reply server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java:249]: If the last character is a newline then in means that there will be a new empty line, so it is not possible in this setup

File: server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+    statementMap.forEach(
+        (tokenName, range) -> {
+          validateTokenName(extendedText, tokenName, range);
+          Location originalLocation = extendedText.mapLocation(range);
+
+          ExtendedTextLine extendedTextLine =
+              extendedText.getLines().get(range.getStart().getLine());
+          String value =
+              extendedTextLine
+                  .subline(range.getStart().getCharacter(), range.getEnd().getCharacter() - 1)
```
Comment (slavek-kucera) [inline server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java]: In this case, I suggest to actually fix the subline to take inclusive/exclusive bounds - it seems to be used only in two places, one of which is test.

File: server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java [hunk 13]
```diff
@@ -373,4 +375,304 @@ void testInsertWithPadding() {
...
+    ExtendedText extendedText = new ExtendedText("MOVE VAR1 TO VAR2", "uri");
+
+    Range range = new Range(new Position(0, 0), new Position(0, 18));
```
Comment (slavek-kucera) [inline server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java:383]: The end position seems to be outside of the extendedText valid range, yet no exception is produced.

File: server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java (hunk 12 above)
Comment (Nurkambay) [reply server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java]: Done

File: server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java (hunk 11 above)
Comment (Nurkambay) [reply server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/TextMapReplacer.java:41]: Done

File: server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java (hunk 13 above)
Comment (Nurkambay) [reply server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java:383]: It is old functionality and there are no checks for this kind of replacement

File: server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java [hunk 14]
```diff
@@ -373,4 +375,304 @@ void testInsertWithPadding() {
...
+    // line 1: DISPLAY
+    Location location = extendedText.mapLocation(new Range(new Position(0, 0), new Position(0, 6)));
+    assertEquals(statementLocation.toString(), location.toString());
+
+    // line 1: VAR2
+    location = extendedText.mapLocation(new Range(new Position(0, 8), new Position(0, 11)));
+    assertEquals(var2Location.toString(), location.toString());
+
+    // line 2: D(ISPLAY)
+    location = extendedText.mapLocation(new Range(new Position(2, 1), new Position(2, 6)));
```
Comment (slavek-kucera) [inline server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java]:
```suggestion
    location = extendedText.mapLocation(new Range(new Position(1, 1), new Position(1, 6)));
```
?

File: server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java [hunk 15]
```diff
@@ -373,4 +375,304 @@ void testInsertWithPadding() {
...
+    assertEquals(statementLocation.toString(), location.toString());
+
+    // line 2: VAR1
+    location = extendedText.mapLocation(new Range(new Position(1, 8), new Position(1, 11)));
```
Comment (slavek-kucera) [inline server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java]:
```suggestion
    location = extendedText.mapLocation(new Range(new Position(1, 8), new Position(1, 12)));
```

File: server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java [hunk 16]
```diff
@@ -373,4 +375,304 @@ void testInsertWithPadding() {
...
+    location = extendedText.mapLocation(new Range(new Position(0, 8), new Position(0, 11)));
+    assertEquals(var1Location.toString(), location.toString());
+
+    // 2nd VAR1
+    location = extendedText.mapLocation(new Range(new Position(0, 22), new Position(0, 25)));
+    assertEquals(var1Location.toString(), location.toString());
+
+    // 2rd VAR1
+    location = extendedText.mapLocation(new Range(new Position(0, 38), new Position(0, 41)));
+    assertEquals(var1Location.toString(), location.toString());
```
Comment (slavek-kucera) [inline server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java:520]:
```suggestion
    // 1st VAR1
    location = extendedText.mapLocation(new Range(new Position(0, 8), new Position(0, 12)));
    assertEquals(var1Location.toString(), location.toString());

    // 2nd VAR1
    location = extendedText.mapLocation(new Range(new Position(0, 22), new Position(0, 26)));
    assertEquals(var1Location.toString(), location.toString());

    // 2rd VAR1
    location = extendedText.mapLocation(new Range(new Position(0, 38), new Position(0, 42)));
    assertEquals(var1Location.toString(), location.toString());
```

File: server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java
Comment (slavek-kucera) [inline server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java:1]: Could you add one more test for the following edge case `STATEMENT AB` -> `STATEMENT {A}\nSTATEMENT {B}`?

Comment (Nurkambay) [reply server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java:1]: Done

File: server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java [hunk 17]
```diff
@@ -373,4 +375,326 @@ void testInsertWithPadding() {
...
+    Range range = new Range(new Position(0, 0), new Position(0, 12));
+    Range statementRange = new Range(new Position(0, 0), new Position(0, 9));
+    extendedText.replace(range, statementRange, statementMap, "STATEMENT {A}\nSTATEMENT {B}");
+
+    Location location =
+        extendedText.mapLocation(new Range(new Position(1, 10), new Position(1, 11)));
+
+    assertEquals("STATEMENT A\n" + "STATEMENT B", extendedText.toString());
+    assertEquals(
+        new Location("uri", new Range(new Position(0, 11), new Position(0, 12))), location);
```
Comment (slavek-kucera) [inline server/common/src/test/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextTest.java]:
```suggestion
    Location locationA =
        extendedText.mapLocation(new Range(new Position(0, 10), new Position(0, 11)));
    Location locationB =
        extendedText.mapLocation(new Range(new Position(1, 10), new Position(1, 11)));

    assertEquals("STATEMENT A\n" + "STATEMENT B", extendedText.toString());
    assertEquals(
        new Location("uri", new Range(new Position(0, 10), new Position(0, 11))), locationA);
    assertEquals(
        new Location("uri", new Range(new Position(0, 11), new Position(0, 12))), locationB);
```

**Review by slavek-kucera — APPROVED** [review]
